exporter/awsemfexporter/metric_translator.go

```diff
@@ -150,215 +200,143 @@ func TranslateCWMetricToEMF(cwMetricLists []*CWMetrics) []*LogEvent {
-	// fields contains metric and dimensions key/value pairs
-	fieldsPairs := make(map[string]interface{})
-	var dimensionArray [][]string
-	// Dimensions Slice
-	var dimensionSlice []string
-	var dimensionKV pdata.StringMap
-	switch metric := dp.(type) {
-	case pdata.IntDataPoint:
-		dimensionKV = metric.LabelsMap()
-	case pdata.DoubleDataPoint:
-		dimensionKV = metric.LabelsMap()
-	}
-
-	dimensionKV.ForEach(func(k string, v pdata.StringValue) {
-		fieldsPairs[k] = v.Value()
-		dimensionSlice = append(dimensionSlice, k)
+// Build CWMetric from DataPoint
+func buildCWMetric(dp DataPoint, pmd *pdata.Metric, namespace string, metricSlice []map[string]string, instrumentationLibName string, config *Config) *CWMetrics {
+	dimensionRollupOption := config.DimensionRollupOption
+	mds := config.MetricDeclarations
+
+	labelsMap := dp.LabelsMap()
+	labelsSlice := make([]string, labelsMap.Len(), labelsMap.Len()+1)
+	// `labels` contains label key/value pairs
+	labels := make(map[string]string, labelsMap.Len()+1)
+	// `fields` contains metric and dimensions key/value pairs
+	fields := make(map[string]interface{}, labelsMap.Len()+2)
+	idx := 0
+	labelsMap.ForEach(func(k string, v pdata.StringValue) {
+		fields[k] = v.Value()
+		labels[k] = v.Value()
+		labelsSlice[idx] = k
+		idx++
 	})
-	// add OTel instrumentation lib name as an additional dimension if it is defined
+
+	// Add OTel instrumentation lib name as an additional dimension if it is defined
 	if instrumentationLibName != noInstrumentationLibraryName {
-		fieldsPairs[OTellibDimensionKey] = instrumentationLibName
-		dimensionArray = append(dimensionArray, append(dimensionSlice, OTellibDimensionKey))
+		labels[OTellibDimensionKey] = instrumentationLibName
+		fields[OTellibDimensionKey] = instrumentationLibName
+	}
+
+	// Create list of dimension sets
+	var dimensionsArray [][][]string
+	if len(mds) > 0 {
+		// Filter metric declarations and map each metric declaration to a list
+		// of dimension sets
+		dimensionsArray = processMetricDeclarations(mds, pmd, labels)
+	} else if instrumentationLibName != noInstrumentationLibraryName {
+		// If no metric declarations defined and OTel instrumentation lib name is defined,
+		// create a single dimension set containing the list of labels +
+		// instrumentationLibName dimension key
+		dimensionsArray = [][][]string{{append(labelsSlice, OTellibDimensionKey)}}
 	} else {
-		dimensionArray = append(dimensionArray, dimensionSlice)
+		// If no metric declarations or OTel instrumentation lib name defined,
+		// create a single dimension set containing only label names
+		dimensionsArray = [][][]string{{labelsSlice}}
+	}
+
+	// Apply single/zero dimension rollup to labels
+	rollupDimensionArray := dimensionRollup(dimensionRollupOption, labelsSlice, instrumentationLibName)
+
+	// Build list of CW Measurements
+	cwMeasurements := make([]CwMeasurement, len(dimensionsArray))
+	for i, dimensions := range dimensionsArray {
```

Comment:
what if dimensionsArray length is zero? Are we also ignoring metrics with rollupDimensionArray here?

Reply:
Yea, that's what is happening in this logic. If `dimensionsArray` is empty, then no `rollupDimensionArray` will be added as a separate metric. Is there a case where `dimensionsArray` is empty, but `rollupDimensionArray` is not? From #7, it seems like this isn't the case because if the label set is empty, then the output `rollupDimensionArray` is also empty.

Reply:
Update from offline discussion: aggregate all dimensions into one large 2D array, including rolled-up dimensions. In this way, the above bug is resolved.